solpolpy/core.py

```diff
@@ -158,6 +168,63 @@ def check_alpha_requirement(path: t.List[str]) -> bool:
     return requires_alpha
 
 
+def generate_imax_matrix(arrayshape) -> np.ndarray:
+    """
+    Define an A matrix with which to convert MZP^ (camera coords) = A x MZP (solar coords)
+
+    Parameters
+    -------
+    arrayshape
+        Defined input WCS array shape for matrix generation
+
+    Returns
+    -------
+    ndarray
+        Output A matrix used in converting between camera coordinates and solar coordinates
+
+    """
+
+    # Ideal MZP wrt Solar North
+    thmzp = [-60, 0, 60] * u.degree
+
+    long_arr, lat_arr = np.meshgrid(np.linspace(-20, 20, arrayshape[0]), np.linspace(-20, 20, arrayshape[1]))
+
+    # Foreshortening (IMAX) effect on polarizer angle
+    phi_m = np.arctan2(np.tan(thmzp[0]) * np.cos(long_arr * u.degree), np.cos(lat_arr * u.degree)).to(u.degree)
+    phi_z = np.arctan2(np.tan(thmzp[1]) * np.cos(long_arr * u.degree), np.cos(lat_arr * u.degree)).to(u.degree)
+    phi_p = np.arctan2(np.tan(thmzp[2]) * np.cos(long_arr * u.degree), np.cos(lat_arr * u.degree)).to(u.degree)
+
+    phi = np.stack([phi_m, phi_z, phi_p])
+
+    # Define the A matrix
+    mat_a = np.empty((arrayshape[0], arrayshape[1], 3, 3))
+
+    for i in range(3):
+        for j in range(3):
+            mat_a[:, :, i, j] = (4 * np.cos(phi[i] - thmzp[j]) ** 2 - 1) / 3
+
+    return mat_a
+
+
+def resolve_imax_effect(input_data: NDCollection) -> NDCollection:
+
+    for i, key in enumerate(input_data.keys()):
+        if i == 0:
+            data_shape = input_data[key].data.shape
+            data_mzp_camera = np.zeros([data_shape[0], data_shape[1], 3, 1])
+        data_mzp_camera[:, :, i, 0] = input_data[key].data
+
+    imax_matrix = generate_imax_matrix(data_shape)
+    imax_matrix_inv = np.linalg.inv(imax_matrix)
```

Comment:
Will this inversion ever fail?

Reply:
For singular matrix case there is a chance. However it didn't fail yet for the WFI cases we tested. Even with some weirdest angles we passed as input (like 90 degrees)

Reply:
https://numpy.org/doc/stable/reference/generated/numpy.linalg.LinAlgError.html